_posts/2016-04-30-akka-basics.md

```diff
@@ -0,0 +1,75 @@
+
+## Fault tolerance
+
+There is not a clear and shared error model in Java, and you can realise that after working in a couple of projects. The origin of that is a confusing abstraction called Exception and its subtypes Checked and Unchecked Exceptions. The only real difference is that checked expections force us to deal with the issue in the direct caller, except for that they don't hint at the strategy that we should apply.
+
+I strongly recommend [this](http://joeduffyblog.com/2016/02/07/the-error-model/) post about it. A powerful idea in the post is that bugs aren't recoverable errors. This quote is contained in that post, [original source](https://wiki.haskell.org/Error_vs._Exception):
+
+> I was involved in the development of a library that was written in C++. One of the developers told me that the developers are divided into the ones who like exceptions and the other ones who prefer return codes. As it seem to me, the friends of return codes won. However, I got the impression that they debated the wrong point: Exceptions and return codes are equally expressive, they should however not be used to describe errors. Actually the return codes contained definitions like ARRAY_INDEX_OUT_OF_RANGE. But I wondered: How shall my function react, when it gets this return code from a subroutine? Shall it send a mail to its programmer? It could return this code to its caller in turn, but it will also not know how to cope with it. Even worse, since I cannot make assumptions about the implementation of a function, I have to expect an ARRAY_INDEX_OUT_OF_RANGE from every subroutine. My conclusion is that ARRAY_INDEX_OUT_OF_RANGE is a (programming) error. It cannot be handled or fixed at runtime, it can only be fixed by its developer. Thus there should be no according return code, but instead there should be asserts.
+
+Akka provides a great way for dealing with failures/errors based on these tenets:
+
+* Single responsibility principle. Failure management is delegated to supervisors, creating focused business actors. We agreed some time ago that lifecycle management, e.g. object creation, should be moved to factories, therefore an object is not responsible for creating itself. Resuming or restarting an object, after some failure happened, is part of that lifecycle management and Akka enforces you to move that responsibility to supervisors.
+
+* Default recovery strategies. It's important to understand which are our options after the system is under some failure condition. Is the DB temporarily down? Has some input, in combination with the existing data, created an undesired state? Is that a bug? Do we need to abandon that particular request or that part of the system is messed up until some patch is applied? Those questions will determine our response to that incident, and Akka provides some [built-in strategies](http://doc.akka.io/docs/akka/2.4.4/scala/fault-tolerance.html#fault-tolerance-scala).
+
```

Comment:
Reading the above paragraphs I'm thinking of design principles like `low coupling` and `high cohesion`, maybe you can say how the Actor / Reactive / Akka system is helping achieve this.